src/schema/jsii2schema.ts

```diff
@@ -312,16 +316,30 @@ function schemaForUnion(type: reflect.TypeReference, ctx: SchemaContext): any {
   return { anyOf };
 }
 
-function schemaForConstructRef(type: reflect.TypeReference) {
+function schemaForConstructRef(type: TypeReference) {
   if (!isConstruct(type)) {
     return undefined;
   }
 
-  return schemaForReferences();
+  return schemaForReferences(type.type);
 }
 
-function schemaForReferences() {
-  return { anyOf: [$ref('FnRef'), $ref('FnGetProp')] };
+function schemaForReferences(type?: reflect.Type) {
+  const anyOf = [];
+  if (type) {
+    const methods = allStaticMethods(type.system).filter((m) =>
+      m.returns.type.type?.extends(type)
+    );
+
+    anyOf.push(
+      ...methods.map((m) => ({
+        [methodFQN(m)]: {},
```

Comment:
I tried to be more strict here, by using `methodSchema()`, but the size of `cdk.schema.json` goes to 24 MB and the build time to 6 min. So I decided to be a bit less strict and we can come back to this later if we decide to.

Reply:
Yeah that's all good. We can optimize the schema later.